---
title: "Integrating proteomics data"
author: "Pratheepa Jeganathan, Susan Holmes"
date: '`r format(Sys.time(), "%d %B, %Y")`'
output: 
  BiocStyle::html_document
params:
  K: 3
  R: 50
---

- Wagner 2018 Cytof
  - 140 breast cancer patients; Of 140, 6 tripple negative (TN)
  - cd45_sce_dropna$Clinical.Subtype == "TN
  - 3 cancer-free
  - 73 protein markers in immune-centric and tumor-centric microenvironment
  
- Keren 2018 Multiplex Ion Bean Imaging (MIBI)
  - Tumor-immune microenvironment in TN patients
  - 41 TN patients
  - 36 proteins
  
We will choose patients with TN and immune cells in both CyTOF and MIBI for the integrative analysis.


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
  message = FALSE,
  warning = FALSE,
  fig.width = 10,
  fig.height = 10)
```


```{r read_arg}
K <- params$K
K
iter <- params$R
iter
```

```{r}
#library(flowCore)
library(SingleCellExperiment)
library(ggplot2)
library(rstan)
#library(scater)
library(plyr)
library(reshape2)
library(readr)
#library(CATALYST)
library(magrittr)
#library(scMerge)
library(MultiAssayExperiment)
library(dplyr)
library(DESeq2)
library(abind)
library(tibble)
library(RColorBrewer)
```


# Mass-Tag CyTOF Breast Cancer Data

- For this analysis we use **cd45.sce**: CD45+ cells from the live cells; downsampled to 426,872 cells (assayed on immune panel) $\times$ 35 proteins

```{r}
# load("../Data/masstagSCE.rda")
# cd45.sce # 38 * 426872 
# rm(epith.sce, livecells.sce, myeloid.sce, tcell.sce)

```

Drop the five subjects without any clinical data. We can use Gender variable to identify those subjects
```{r}
# unique(colData(cd45.sce)$patient_id.y[which(is.na(colData(cd45.sce)$Gender))])

# cd45_to_keep <- which(!is.na(colData(cd45.sce)$Gender))
# cd45.sce_dropna <- cd45.sce[,cd45_to_keep]
# cd45.sce_dropna # 38 * 420685
# # To verify there are no (true) NA's left:
# sum(is.na(rowData(cd45.sce_dropna)))
# rm(cd45.sce, cd45_to_keep)
```


```{r}
#saveRDS(cd45.sce_dropna, "../Data/cd45_sce_dropna.rds")
cd45_sce_dropna <- readRDS("../Data/cd45_sce_dropna.rds")
```

# Keren et al., MIBI-TOF Breast Cancer Data 


- All the patients were TN 
- The size-normalized raw intensity values are then arcsinh transformed and standardized across the markers. 

```{r}
load('../Data/mibiSCE.rda')
mibi.sce
```

Rows correspond to channels and columns correspond to cells.
We can see all of the channels that were collected in this experiment:

```{r}
rownames(mibi.sce)
```



The 38 proteins can be easily identified by using the binary attribute is_protein from rowData:

```{r}
proteins <- rowData(mibi.sce)$is_protein == 1
sum(proteins)
mibi.sce_proteins <- mibi.sce[which(proteins),]
mibi.sce_proteins
rownames(mibi.sce_proteins)

rm(mibi.sce)
```


Cell type information is availble in the columns *tumor_group* and *immune_group* 

51% of cells were Keratin positive tumor cells and 41% of cells were immune cells.

```{r}
round(table(mibi.sce_proteins$tumor_group)/ncol(mibi.sce_proteins),2)
```


Among the immune cell population, macrophages and CD8, CD4+ T-cells and other immune cells were identified.  10% of all cells assayed were macrophages

```{r}
# Immune Cells
round(table(mibi.sce_proteins$immune_group)/ncol(mibi.sce_proteins),2)
```


These fields can be used to subset by cell and/or sample characteristic. E.g., below we subset only for the 83,336 cells identified as immune.

```{r}
immune_inds <- which(mibi.sce_proteins$tumor_group == "Immune")
mibi.sce_immune <- mibi.sce_proteins[,immune_inds]
mibi.sce_immune

mibi_sce_immune <- mibi.sce_immune

rm(mibi.sce_proteins, mibi.sce_immune)
```



There are three proteins without gene symbol in cd45_sce_dropna so we drop them
```{r}
Proteins_in_cd45 <- !(rowData(cd45_sce_dropna)$hgnc_symbol == "na")

cd45_sce_dropna <- cd45_sce_dropna[which(Proteins_in_cd45),]
cd45_sce_dropna
```



Proteins in both CD45 cells from the live cells (assayed on immune panel) and donor immune cells
```{r}
sum(rownames(cd45_sce_dropna) %in% rownames(mibi_sce_immune))
rownames(cd45_sce_dropna)[which(rownames(cd45_sce_dropna) %in% rownames(mibi_sce_immune))]
```



## Merge two SingleCellExperiment objects and make mae

```{r}
# 148 patients
se_cd45 <- SummarizedExperiment(
  assays = list(exprs = assay(cd45_sce_dropna)),
  colData = colData(cd45_sce_dropna),
  rowData = rowData(cd45_sce_dropna)
)

# Subset CD45 cells from the live cells from TN patients

se_cd45 <- se_cd45[, colData(se_cd45)$Clinical.Subtype == "TN"]
se_cd45 <- se_cd45[, colSums(assay(se_cd45)) > 0]

se_mibi <- SummarizedExperiment(
  assays = list(exprs = assay(mibi_sce_immune)),
  colData = colData(mibi_sce_immune),
  rowData = rowData(mibi_sce_immune)
)

se_mibi <- se_mibi[, colSums(assay(se_mibi)) > 0]

```


```{r}
names(colData(se_cd45))[1:3] <- c("SampleID", "Condition", "PatientID")
names(colData(se_mibi))[c(1,9)] <- c("SampleID", "PatientID")

colData(se_cd45)$SampleID <- paste0(seq(1: length(colData(se_cd45)$SampleID)), "_", colData(se_cd45)$PatientID)

colData(se_mibi)$SampleID <- paste0(seq(1: length(colData(se_mibi)$SampleID)), "_", colData(se_mibi)$PatientID)

rownames(colData(se_cd45)) <- colData(se_cd45)$SampleID
rownames(colData(se_mibi)) <- colData(se_mibi)$SampleID
```

Subset for the initial analysis
```{r}
# se_cd45 <- se_cd45[, colData(se_cd45)$PatientID == "BB028"]
# colData(se_mibi)$PatientID <- factor(colData(se_mibi)$PatientID)
# se_mibi <- se_mibi[, colData(se_mibi)$PatientID == "30847"]
se_cd45 <- se_cd45[, sample(1:dim(se_cd45)[2], 50)]
se_mibi <- se_mibi[, sample(1:dim(se_mibi)[2], 50)]
```

```{r}
cd45_mibi <- list("cd45" = se_cd45, "mibi" = se_mibi)

cd45_map <- data.frame(primary = as.character(colData(se_cd45)$SampleID),
                     colname = as.character(colData(se_cd45)$SampleID), 
                     stringsAsFactors = FALSE)

mibi_map <- data.frame(primary = as.character(colData(se_mibi)$SampleID),
                     colname = as.character(colData(se_mibi)$SampleID), 
                     stringsAsFactors = FALSE)

listmap <- list(cd45_map, mibi_map)
names(listmap) <- names(cd45_mibi)

dfmap <- listToMap(listmap = listmap)

sam_se_cd45 <- colData(se_cd45) %>% data.frame() %>% as.tibble()
sam_se_mibi <- colData(se_mibi) %>% data.frame() %>% as.tibble()
map_df <- full_join(sam_se_cd45, sam_se_mibi, by = "SampleID")
map_df %<>% as.data.frame
rownames(map_df) <- map_df$SampleID 

mae <- MultiAssayExperiment(experiments = cd45_mibi,
                             colData = map_df, 
                             sampleMap = dfmap)
```


Note: we need to do inverse of arcsinh. Then, do median-of-ratios, make integer values.

```{r}
trans_func <- function(x){
  asinh(x/5)
  }
inv_func <- function(x) {
  5*sinh(x)
}
inv_func2 <- function(x) {
  sinh(x)
}
# what normalization for cd45_sce_dropna expression
# asinh transformation with cofactor 5
assay(mae[["cd45"]]) <- inv_func(assay(mae[["cd45"]]))
# what normalization mibi_sce_immune
# cell size normalized, archsinh, standardized across the markers
# we couldn't guess the protein marker mean and standard deviation. We will  
rw_mins <- rowMins(assay(mae[["mibi"]]))*(-1)

assay(mae[["mibi"]]) <-  assay(mae[["mibi"]]) + rw_mins

assay(mae[["mibi"]]) <- inv_func2(assay(mae[["mibi"]]))

names(rowData(mae[["mibi"]]))[1] <- "marker_name"
```

Remove features with all zero values
```{r}
for(i in 1:length(mae)){
  not_all_zeros <- apply(assay(mae, i), 1, function(x){sum(x > 0) > 1})
  names_not_all_zeros <- names(not_all_zeros)[which(not_all_zeros)]
  row_names <- rownames(mae)
  row_names[[names(mae)[i]]] <- names_not_all_zeros
  mae = mae[row_names, , ]
}
```


# Converting each scale to have the same lower and upper levels

$y = \left(\dfrac{x-x_{\text{min}}}{x_{\text{range}}}\right) \times u$, where $u$ is the upper limit of the rescaled variable.

We will scale cd45 and mibi to upper limit of cd45
```{r}
u <- max(assay(mae[["cd45"]])) 
x_mibi <- assay(mae[["mibi"]]) 
y_mibi <- apply(x_mibi, 1, function(x){
  xmin <- min(x)
  xrange <- max(x) - min(x)
  y <- (x-xmin)/xrange*u
  return(y)
})
y_mibi <- t(y_mibi)

#y_mibi[is.na(y_mibi)] <- 0
assay(mae[["mibi"]]) <- y_mibi

```


# Topic modeling

```{r}
x1 <- assay(mae[["cd45"]]) %>% t() %>% as.tibble()
x2 <- assay(mae[["mibi"]]) %>% t() %>% as.tibble()
x <-  full_join(x1, x2)
common_markers <- colnames(x1)[colnames(x1) %in% colnames(x2)]
common_markers
rm(x1,x2) 

# some features are not recorded in one domain
x[is.na(x)] <- 0

ot <- t(x)
geo_mean <- function(x) {
        if(all(x == 0)){
            val <- 0
        }else{
            val <- exp(sum(log(x[x > 0]))/length(x))
        }
        return(val)
    }

geom_mean_row <- apply(ot, 1, FUN = geo_mean)


dj <- estimateSizeFactorsForMatrix(ot, median, geoMeans = geom_mean_row)

ot <- t(t(ot)/dj)

x <- t(ot)
rm(ot)
x <- apply(x , c(1,2), function(y){as.integer(y)})
x <- as.matrix(x)
```

```{r}
# theta[d] ~ dirichlet(alpha), alpha pseudocount for each topic
# beta[k] ~ dirichlet(gamma), gamma pseudocount for each ASV in each topic
stan.data <- list(K = K, 
  V = ncol(x), 
  D = nrow(x), 
  n = x, 
  alpha = rep(.8, K), 
  gamma = rep(.5, ncol(x))
)
```

```{r}
fileN <- paste0("../Results/proteomics_filtered_K_",K,"_ite_",iter,".RData")
fileN
```


```{r}
t1 <- proc.time()
stan.fit <- stan(file = "./lda.stan", 
  data = stan.data, 
  iter = iter, 
  chains = 4, 
  sample_file = NULL,
  diagnostic_file = NULL,
  cores = 4,
  control = list(adapt_delta = 0.9),
  save_dso = TRUE,
  algorithm = "NUTS")
proc.time() - t1

save(stan.fit, file = fileN)
```

```{r}
load(file = fileN)
```

# Stan fit

## summary of the log-posterior
```{r}
# head(rstan::extract(stan.fit)$"lp__")
# 
# tail(rstan::extract(stan.fit)$"lp__")

df <- data.frame(lp = rstan::extract(stan.fit)$"lp__", Chain = paste0("Chain ", rep(seq_len(4), each = (iter/2))))

ggplot(data = df) +
  geom_histogram(aes(x = lp, col = Chain, fill = Chain))
```

## Posterior mean, Monte Carlo standard error, posterior standard deviation
```{r}
head(summary(stan.fit, pars = c("theta"), probs = c(0.1, 0.9))$summary)

head(summary(stan.fit, pars = c("beta"), probs = c(0.1, 0.9))$summary)

# plot R hat of theta and beta
hist(summary(stan.fit, pars = c("theta", "beta"), probs = c(0.1, 0.9))$summary[,7])

```

## Sampler diagnostics
```{r}
sampler_params <- get_sampler_params(stan.fit, inc_warmup = FALSE)
colnames(sampler_params[[1]])

mean_accept_stat_by_chain <- sapply(sampler_params, function(x) mean(x[, "accept_stat__"]))
mean_accept_stat_by_chain

max_treedepth_by_chain <- sapply(sampler_params, function(x) max(x[, "treedepth__"]))
max_treedepth_by_chain
```

# Visualization 

## Extract posterior samples
```{r}
samples <- rstan::extract(stan.fit, permuted = TRUE, inc_warmup = FALSE, include = TRUE)# samples is a list
```

## Alignment
- Create a Topic $*$ Chain matrix
```{r}
source("../R_BANF/alignmentMatrix.R")
source("../R_BANF/thetaAligned.R")
theta <- samples$theta 
aligned <- alignmentMatrix(theta, mae, K, iter = iter, chain = 4)
theta_aligned <- thetaAligned(theta, K, aligned, iter = iter, chain = 4)
dimnames(theta_aligned)[[2]] <- colnames(mae) %>% unlist() %>% as.character()
dimnames(theta_aligned)[[3]] <- c(paste0("Topic_", seq(1,K)))
```


```{r}
# array to a dataframe
theta_all <- melt(theta_aligned)
colnames(theta_all) <- c("iteration", "Sample", "Topic", "topic.dis")
theta_all$Chain <- paste0("Chain ", rep(seq(1, 4), each = (iter/2)))

sam <- colData(mae) %>% data.frame()
theta_all$Sample <- as.character(theta_all$Sample)
theta_all <- left_join(theta_all, sam, by =c("Sample"= "SampleID"))
theta_all$Chain <- factor(theta_all$Chain)
theta_all$Topic <- factor(theta_all$Topic)
theta_all$immune_group[which(is.na(theta_all$immune_group))] <- "Unknown"
theta_all$immune_group <- factor(theta_all$immune_group)
```

## Plot the topic distribution in each chain
```{r}
p_chain <- ggplot(data = theta_all) + 
  geom_boxplot(aes(x = Sample, y = topic.dis, 
                   color = Topic, group = Sample)) + 
  facet_grid(Topic ~ Chain, scales = "free_x")+
  ylab(bquote(theta[k])) + 
  ggtitle("Topic distribution for CD45 and MIBI-immune cells") + 
  xlab("Cells") +
  theme(plot.title = element_text(hjust = 0.5, size = 20), 
    strip.text.y= element_text(size = 15, angle = 90), 
    strip.text.x = element_text(size = 15),
    axis.text.x = element_blank(), 
    axis.text.y=element_text(size = 15),
    axis.title = element_text(size = 15),
    legend.position = "none")  
p_chain

```


## Plot the topic distribution in all chains
```{r}
p <- ggplot(data = theta_all) + 
  geom_boxplot(aes(x = Sample, y = topic.dis, 
                   color = Topic, group = Sample)) + 
  facet_grid(Topic ~ immune_group, scales = "free_x")+
  ylab(bquote(theta[k])) + 
  ggtitle("Topic distribution for CD45 and MIBI-immune cells") + 
  xlab("Cells") +
  theme(plot.title = element_text(hjust = 0.5, size = 20), 
    strip.text.y= element_text(size = 15, angle = 90), 
    strip.text.x = element_text(size = 15),
    axis.text.x = element_blank(), 
    axis.text.y=element_text(size = 15),
    axis.title = element_text(size = 15),
    legend.position = "none") 
p
```



## Marker distribution
```{r}
source("../R_BANF/betaAligned.R")
beta <- samples$beta # an array (iterations *topic * ASV)
beta_aligned <- betaAligned(beta, K, aligned, iter = iter, chain = 4) # an array (iterations *topic * ASV)

# g(beta_k) transformation on beta_k
beta_logit <- beta_aligned
rm(beta_aligned)
for (i in seq_len(nrow(beta_logit))) {
  for (k in seq_len(K)) {
    beta_logit[i, k, ] <- log(beta_logit[i, k, ])
    beta_logit[i, k, ] <- beta_logit[i, k, ] - mean(beta_logit[i, k, ])
  }
}

# array to data frame
beta_hat <- beta_logit %>%
  melt(varnames = c("iterations", "topic", "marker_ix"), value.name = "beta_logit") %>%
  as_tibble()
beta_hat$marker <- colnames(x)[beta_hat$marker_ix]
rm(beta_logit)

# join rowData with beta_hat
marker_info <- full_join(rowData(mae[["cd45"]]) %>% data.frame(), rowData(mae[["mibi"]]) %>% data.frame())
marker_info$marker <- marker_info$marker_name
# If we use a taxanomy level with NA, we can replace the taxanomy level with one level before this level
#marker_info$Class[which(is.na(marker_info$Class))] <- marker_info$Phylum[which(is.na(marker_info$Class))]

beta_hat <- beta_hat %>%
  left_join(marker_info, by = "marker") %>% mutate(topic = paste("Topic", topic))

# sort ASV by the number of ASV per Class (there is no NA, if there is NA, we need to replace by some character to make sure that NA does not dominate on sort)
sorted_marker <- names(sort(table(beta_hat$marker_name), decreasing = TRUE))
beta_hat$marker_name <- factor(beta_hat$marker_name)
beta_hat$marker_name <- factor(beta_hat$marker_name, levels = sorted_marker)
beta_hat$marker <- factor(beta_hat$marker)
beta_hat$marker_ix <- factor(beta_hat$marker_ix)
beta_hat$topic <- factor(beta_hat$topic)


beta_summary <- beta_hat %>% 
  dplyr::group_by(marker_ix, topic) %>% 
  dplyr::summarise(
    marker = marker[1],
    beta_median = median(beta_logit),
    marker_name = marker_name[1],
    hgnc_symbol = hgnc_symbol[1],
    beta_upper = quantile(beta_logit, 0.975),
    beta_lower = quantile(beta_logit, 0.025)
  )

# we plot only 20 marker_name and merge other marker_name to "other"
levels(beta_summary$marker_name) <- c(levels(beta_summary$marker_name), "other")
choose_marker <- c(unique(beta_summary$marker_name[order(beta_summary$beta_median, decreasing = TRUE)])[1:20] %>% as.character(), common_markers)
choose_marker <- unique(choose_marker)
beta_summary$marker_name[!(beta_summary$marker_name %in% choose_marker)] <- "other"
# beta_summary$marker_name[!(beta_summary$marker_name %in% levels(beta_summary$marker_name)[1:30])] <- "other"
beta_summary$marker_name <- droplevels(beta_summary$marker_name)

# we plot all 64 markers (the order was determined by total sum of reads per marker) and color by hgnc_symbol
x_order <- names(sort(colSums(x), decreasing = TRUE))
beta_subset <- beta_summary %>%
  dplyr::filter(marker_name %in% x_order)
beta_subset$marker_ix <- rep(seq_len(nrow(beta_subset) / K), each = K)

```


```{r}
getPalette <- colorRampPalette(brewer.pal(9, "Set1"))
```

```{r}
p <- ggplot(beta_subset) +
  geom_hline(yintercept = 0, alpha = 0.4, size = 0.5, col = "#999999") +
  geom_point(aes(x = marker_ix, y = beta_median, col = marker_name), size = 1) +
  geom_errorbar(aes(x = marker_ix, alpha = beta_upper, ymax = beta_upper, ymin = beta_lower, col = marker_name), size = 0.4) + 
   facet_grid(topic ~ .) + 
  scale_color_manual(values = getPalette(length(choose_marker))) +
  scale_alpha(range = c(0.01, 1), breaks = c(1, 2, 3), guide = FALSE) + ## larger values have darker CIs
  scale_x_continuous(expand = c(0, 0)) +
  guides(color = guide_legend(override.aes = list(alpha = 1, size = 2))) +
  labs(x = "marker name", y = expression(paste("g(", beta[k], ")")), col = "marker") +
  theme(panel.border = element_rect(fill = "transparent", size = 0.75),
    axis.text.x = element_blank(),
    strip.text.x = element_blank(),
    legend.position = "bottom")

p
```

# Without including the spatial x-y coordinate data, how well can we predict cell co-location?

```{r}
theta_summary <- theta_all %>%
  dplyr::group_by(Sample, Topic) %>%
  dplyr::summarise(
    median_theta = median(topic.dis)
  )

theta_summary_assign <- theta_summary %>%
  dplyr::group_by(Sample) %>%
  dplyr::summarise(
    value = max(median_theta),
    Topic = Topic[median_theta == max(median_theta)]
  )

```

The cells have large proportion of topic 1
```{r}
theta_summary_assign$Sample[theta_summary_assign$Topic == "Topic_1"]
colData(mae[, theta_summary_assign$Sample[theta_summary_assign$Topic == "Topic_1"], ])
```


```{r}
colData(mae[, theta_summary_assign$Sample[theta_summary_assign$Topic == "Topic_2"], ])
```

```{r}
colData(mae[, theta_summary_assign$Sample[theta_summary_assign$Topic == "Topic_3"], ])
```
